common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +173,446 @@
+                return getExtensionInstance(cachedExtensionDefinition, loader, argTypes, args);
+            } catch (Throwable e) {
+                if (e instanceof EnhancedServiceNotFoundException) {
+                    throw (EnhancedServiceNotFoundException)e;
+                } else {
+                    throw new EnhancedServiceNotFoundException(
+                            "not found service provider for : " + type.getName() + " caused by " + ExceptionUtils
+                                    .getFullStackTrace(e));
+                }
             }
-            return result;
-        } catch (Throwable e) {
-            if (e instanceof EnhancedServiceNotFoundException) {
-                throw (EnhancedServiceNotFoundException)e;
+        }
+
+        private S getExtensionInstance(ExtensionDefinition definition, ClassLoader loader, Class[] argTypes,
+                                       Object[] args) {
+            if (definition == null) {
+                throw new EnhancedServiceNotFoundException("not found service provider for : " + type.getName());
+            }
+            if (Scope.SINGLETON.equals(definition.getScope())) {
+                Holder<Object> holder = definitionToInstanceMap.get(definition);
+                if (holder == null) {
+                    definitionToInstanceMap.putIfAbsent(definition, new Holder<>());
+                    holder = definitionToInstanceMap.get(definition);
+                }
+                Object instance = holder.get();
+                if (instance == null) {
+                    synchronized (holder) {
+                        instance = holder.get();
+                        if (instance == null) {
+                            instance = createNewExtension(definition, loader, argTypes, args);
+                            holder.set(instance);
+                        }
+                    }
+                }
+                return (S)instance;
             } else {
-                throw new EnhancedServiceNotFoundException(
-                    "not found service provider for : " + service.getName() + " caused by " + ExceptionUtils
-                        .getFullStackTrace(e));
+                return createNewExtension(definition, loader, argTypes, args);
             }
         }
-    }
 
-    @SuppressWarnings("rawtypes")
-    private static <S> List<Class> findAllExtensionClass(Class<S> service, String activateName, ClassLoader loader) {
-        List<Class> extensions = new ArrayList<>();
-        try {
-            loadFile(service, SERVICES_DIRECTORY, loader, extensions);
-            loadFile(service, SEATA_DIRECTORY, loader, extensions);
-        } catch (IOException e) {
-            throw new EnhancedServiceNotFoundException(e);
+        private S createNewExtension(ExtensionDefinition definition, ClassLoader loader, Class[] argTypes, Object[] args) {
+            loadAllExtensionClass(loader);
```

Comment:
if necessary？

Reply:
I checked all the call stacks of createNewExtension(), found that when createNew() is executed, all extension classes have been loaded, so it's unnecessary.I will remove it.